components/camel-dynamic-router/src/main/java/org/apache/camel/component/dynamicrouter/DynamicRouterEndpoint.java

```diff
@@ -140,30 +149,87 @@ protected void doInit() throws Exception {
+                            .getInstance(component);
+                    processor.setConfiguration(configuration);
                     component.setControlChannelProcessor(processor);
                 }
             } catch (Exception e) {
                 throw new IllegalStateException("Could not create Dynamic Router endpoint", e);
             }
         } else {
-            final DynamicRouterProcessor processor = processorFactorySupplier.get()
-                    .getInstance("dynamicRouterProcessor-" + configuration.getChannel(), getCamelContext(),
-                            configuration.getRecipientMode(), configuration.isWarnDroppedMessage(),
-                            filterProcessorFactorySupplier);
-            ServiceHelper.startService(processor);
+            CamelContext camelContext = getCamelContext();
+            String routeId = configuration.getRouteId();
+            long timeout = configuration.getTimeout();
+            ErrorHandler errorHandler = new NoErrorHandler(null);
+            if (producerCache == null) {
+                producerCache = new DefaultProducerCache(this, camelContext, 1000);
+            }
+            ExecutorService aggregateExecutorService = camelContext.getExecutorServiceManager()
+                    .newScheduledThreadPool(this, "DynamicRouter-AggregateTask", 0);
+            if (timeout > 0) {
+                // use a cached thread pool so we each on-the-fly task has a dedicated thread to process completions as they come in
+                aggregateExecutorService = camelContext.getExecutorServiceManager()
+                        .newScheduledThreadPool(this, "DynamicRouter-AggregateTask", 0);
+            }
+            AggregationStrategy aggregationStrategy = determineAggregationStrategy(camelContext);
+            DynamicRouterMulticastProcessor processor = processorFactorySupplier.get()
+                    .getInstance("DynamicRouterMulticastProcessor-" + configuration.getChannel(), camelContext, null,
+                            configuration.getRecipientMode(),
+                            configuration.isWarnDroppedMessage(), filterProcessorFactorySupplier, producerCache,
+                            aggregationStrategy, configuration.isParallelProcessing(),
+                            determineExecutorService(camelContext), configuration.isShutdownExecutorService(),
+                            configuration.isStreaming(), configuration.isStopOnException(), timeout,
+                            determineOnPrepare(camelContext), configuration.isShareUnitOfWork(),
+                            configuration.isParallelAggregate());
+            processor.setErrorHandler(errorHandler);
+            processor.setAggregateExecutorService(aggregateExecutorService);
+            processor.setIgnoreInvalidEndpoints(configuration.isIgnoreInvalidEndpoints());
+            processor.setId(getId());
+            processor.setRouteId(routeId);
```

Comment:
I'd probably split this part outside of the init method. I've been broken these large methods in core for a while because: 1. they are hard to read and maintain and 2. they affect JIT compiler ability to inline code when needed. 
Even though this is not on the hot path, I would still like to keep the code easier to read/maintain.

Reply:
Thanks.  I have no problem with this, so I will make the change that you requested.